### PR TITLE
fix(sops): make encrypt idempotent and validate explicit config path

### PR DESCRIPTION
**Pull request by jainal09**
Fixes two encryption/SOPS robustness findings from the reliability audit in #413 (cluster G — encryption/SOPS robustness).

### [HIGH] `envdrift encrypt --backend sops` is not idempotent — re-running on an already-encrypted file exits 1
`src/envdrift/encryption/sops.py:203-297` · `src/envdrift/cli_commands/encryption.py:286-298`

SOPS refuses to re-encrypt a file that already carries a top-level `sops` metadata block and exits non-zero, so `SOPSEncryptionBackend.encrypt()` raised `EncryptionBackendError` → exit 1 on **any** second run (a pre-commit hook firing twice, a CI re-run, a documented re-run). Unlike the dotenvx path there was no idempotency.

**Fix:** `encrypt()` now short-circuits an already-encrypted **in-place** target (detected via `has_encrypted_header`) to a clean no-op success — `"<file> is already encrypted (no change)"`, exit 0 — mirroring dotenvx's `no change`. An explicit `output_file` is a distinct target and still gets written.

### [MEDIUM] Explicit but missing SOPS `--sops-config` / `sops_config_file` is silently ignored
`src/envdrift/encryption/sops.py:182-184` · encrypt/decrypt flows `cli_commands/encryption.py:586-590, 642-643`

`_run` only appended `--config <path>` when the file existed, so a wrong `--sops-config` (or `sops_config_file` in `envdrift.toml`) was silently dropped. SOPS then fell back to an ambient `.sops.yaml` (wrong keys, exit 0 — a data-integrity hazard) or failed with a generic "no creation rules" error, never telling the user their explicit path did not exist.

**Fix:** `_run` now raises `EncryptionBackendError('SOPS config file not found: <path>')` when an explicit `config_file` does not exist. `self._config_file` is only ever set from an explicit CLI `--sops-config` or TOML `sops_config_file` (never auto-discovery), so a missing one is always an explicit-but-wrong path. Applies to both the encrypt and decrypt flows. Fully-implicit auto-discovery (no `config_file`) still tolerates absence.

## Regression tests (verified: fail on unfixed code, pass with the fix)

`tests/unit/test_encryption_sops.py`
- `test_encrypt_already_encrypted_is_idempotent_noop` — re-encrypting an already-encrypted file is a clean no-op, `_run`/sops never invoked.
- `test_encrypt_plaintext_still_runs_sops` — short-circuit does NOT fire for plaintext.
- `test_run_explicit_missing_config_raises` — explicit missing config raises `SOPS config file not found`, never reaching `subprocess.run`.
- `test_run_explicit_present_config_is_used` / `test_run_no_config_does_not_add_flag` — present config still passes `--config`; no config omits it without error.

`tests/integration/test_sops_backend_e2e.py` (real `sops` + `age` binary, skip-gated)
- `test_encrypt_twice_is_idempotent_clean_noop` — encrypt twice → clean no-op, byte-identical ciphertext, still round-trips.
- `test_encrypt_explicit_missing_config_raises_clear_error` — explicit-missing config raises and leaves the plaintext untouched (never encrypted with ambient keys).

## Gates
`ruff check` / `ruff format --check` / `pyrefly check` (incl. test files via sibling run) / `bandit` all pass. `sops.py` coverage 92%. Full non-integration suite: 2290 passed. SOPS integration suite (real binary): 23 passed. Docs updated in `docs/cli/encrypt.md`; `make lint-docs` clean.

Refs #413

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<h3>Greptile Summary</h3>

This PR fixes two correctness issues in the SOPS encryption backend: making `encrypt()` idempotent for already-encrypted in-place targets, and converting a silent config-path fallback into a loud, early failure when an explicit `--sops-config` path does not exist.

- **Idempotency**: `_already_encrypted_noop()` reads the target file and short-circuits to success when the genuine line-anchored SOPS metadata block is present, avoiding SOPS's own non-zero exit on re-encrypt. An explicit `output_file` bypasses the short-circuit since it must always be written. The detection uses `_has_sops_metadata_block()` rather than the looser `ENC[AES256_GCM,` substring check, preventing false no-ops on files that merely comment about ciphertext format.
- **Explicit config validation**: `_config_args()` now raises `EncryptionBackendError` immediately when an explicit config path does not exist, eliminating the silent ambient `.sops.yaml` fallback. Critically, this check is invoked _before_ the idempotency short-circuit in `encrypt()` (addressed in a follow-up commit), so a misconfigured `--sops-config` is always surfaced even when the target is already encrypted.

<h3>Confidence Score: 5/5</h3>

Safe to merge — both fixes are scoped to the SOPS backend, the critical ordering concern raised in review was addressed before this version, and test coverage is comprehensive.

The idempotency short-circuit keys off the genuine line-anchored SOPS metadata block rather than the looser ENC[AES256_GCM, substring, so plaintext files that merely reference ciphertext format are correctly passed through to SOPS. The config-validation guard runs before the idempotency bypass in encrypt(), and decrypt() reaches _run() unconditionally so the config check there is preserved. No pre-existing paths are regressed.

No files require special attention.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/envdrift/encryption/sops.py | Core change: adds _resolve_config_path, _config_args, _already_encrypted_noop, _encrypt_key_args, and _encrypt_output_args helpers; refactors encrypt() to call _config_args before the idempotency check; extracts _has_sops_metadata_block from has_encrypted_header. Logic is correct and the ordering fix (config guard before idempotency short-circuit) is in place. |
| tests/unit/test_encryption_sops.py | Adds 10 new unit tests covering idempotency, plaintext-with-ENC-substring edge case, missing/present/relative/no-config scenarios, and the config-before-idempotency ordering. Coverage is thorough with no gaps in the happy/unhappy paths. |
| tests/integration/test_sops_backend_e2e.py | Adds 3 real-binary integration tests: idempotent double-encrypt, plaintext-with-ENC-substring not mistaken as encrypted, and explicit-missing-config raising clearly. All gated behind _require_sops(). Well-structured, assertions cover byte-for-byte ciphertext stability and round-trip integrity. |
| docs/cli/encrypt.md | Adds two new items to the SOPS behaviour list documenting idempotency and explicit-config validation. Accurate and consistent with implementation. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["encrypt(env_file, **kwargs)"] --> B{file exists?}
    B -- No --> Z1[return failure: file not found]
    B -- Yes --> C{is_installed?}
    C -- No --> Z2[raise EncryptionNotFoundError]
    C -- Yes --> D["_config_args(cwd) — validate explicit config path"]
    D -- config missing --> Z3[raise EncryptionBackendError: SOPS config file not found]
    D -- config OK / no config --> E["_already_encrypted_noop(env_file, kwargs)"]
    E -- output_file set or read error or no metadata block --> F["_encrypt_output_args()"]
    E -- in-place and metadata block found --> G[return success: already encrypted no change]
    F -- in_place=False no output_file --> Z4[return failure: ciphertext would be discarded]
    F -- output_file or in_place --> H["_run(args, env, cwd)"]
    H --> I["_config_args(cwd) + subprocess SOPS call"]
    I -- returncode != 0 --> Z5[raise EncryptionBackendError: SOPS encryption failed]
    I -- returncode 0 --> J[return success: Encrypted path]
```

<sub>Reviews (4): Last reviewed commit: ["Merge branch &#39;main&#39; into fix/sops-idempo..."](https://github.com/jainal09/envdrift/commit/7bab16e9fb306ae5d0f8797b940bdff59f111f6c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=36086983)</sub>

### CODE REVIEW

**Comment by jainal09** [comment]
Bot review findings on the latest PR head (`38d1312`):

1. High: `src/envdrift/encryption/sops.py:263` treats any raw `ENC[AES256_GCM,` substring as proof that a file is already SOPS-encrypted. `has_encrypted_header()` returns true even for comments or prose, so a plaintext `.env` containing `# SOPS values look like ENC[AES256_GCM,...]` plus `DB_PASSWORD=hunter2` reports `already encrypted (no change)` and never invokes SOPS, leaving the secret unencrypted. Please require real SOPS metadata or a structural non-comment encrypted value before no-oping, and add a regression test for comments/prose containing this marker.

2. Medium: the same early return runs before explicit config-file validation in `_run()`. If `config_file` points at a missing path and the input already has SOPS markers, `encrypt()` returns success with `already encrypted (no change)` instead of surfacing `SOPS config file not found`, so a bad `--sops-config` or `sops_config_file` is silently masked. Please validate `self._config_file` before this early return, or factor the validation into a helper used before any SOPS operation can short-circuit.

3. Medium: `src/envdrift/encryption/sops.py:190` checks relative `config_file` paths against the Python process cwd, but `_run()` executes SOPS with a caller-provided `cwd`. A valid call such as `SOPSEncryptionBackend(config_file=".sops.yaml")._run(..., cwd=workspace)` now fails with `SOPS config file not found: .sops.yaml` even when `workspace/.sops.yaml` exists. Please resolve relative config paths against `cwd` for both validation and the `--config` argument, or normalize them before constructing the backend.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>

**Comment by jainal09** [comment]
Bot review summary: reviewed latest head `38d1312` for PR #422, including the SOPS backend idempotency/config changes, linked #413 context, visible comments, and CI status.

Posted 3 findings in this review pass: plaintext files can be skipped if they merely mention the SOPS ciphertext marker, a missing explicit SOPS config is bypassed by the idempotency early return, and relative config validation uses process cwd instead of the `_run(cwd=...)` context. Targeted tests passed in the isolated review, but ad hoc repros covered these missing cases.

<sub>Bot summary posted by Codex via `gh` using Jainal's account.</sub>

**Comment by jainal09** [comment]
Bot CI status update: after re-polling `gh pr checks` for latest head `38d1312`, the GitHub Actions code/test jobs are passing, but these checks are currently failing: `CodeScene Code Health Review (main)`, `codecov/patch`, and `codecov/project`. Cloud Provider Tests are skipped.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>

File: src/envdrift/encryption/sops.py [hunk 1]
```diff
@@ -239,6 +247,26 @@ def encrypt(
...
+            try:
+                existing = env_file.read_text()
+            except (OSError, UnicodeDecodeError):
+                existing = ""
+            if existing and self.has_encrypted_header(existing):
+                return EncryptionResult(
+                    success=True,
+                    message=f"{env_file} is already encrypted (no change)",
+                    file_path=env_file,
+                )
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/encryption/sops.py]: Fixed in d1dfb25. The explicit-config validation (`_config_args`) now runs *before* the idempotency short-circuit in `encrypt()`, so a missing `--sops-config` is surfaced as `SOPS config file not found` even when the target is already encrypted. Regression test: `test_encrypt_missing_config_raises_before_idempotency_noop` (unit) plus the real-binary `test_encrypt_explicit_missing_config_raises_clear_error`.

File: src/envdrift/encryption/sops.py [hunk 2]
```diff
@@ -239,6 +247,26 @@ def encrypt(
...
+        # path and treat an already-encrypted in-place target as a clean no-op.
+        # Skipped only for in-place encryption (no output_file): an explicit
+        # output_file is a distinct target and must still be written.
+        in_place_target = not kwargs.get("output_file")
+        if in_place_target:
+            try:
+                existing = env_file.read_text()
+            except (OSError, UnicodeDecodeError):
+                existing = ""
+            if existing and self.has_encrypted_header(existing):
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/encryption/sops.py]: Fixed in d1dfb25. The idempotency no-op now keys off the genuine line-anchored SOPS metadata block (`_has_sops_metadata_block`, using `SOPS_METADATA_PATTERNS`) instead of `has_encrypted_header`, so a bare `ENC[AES256_GCM,` substring in plaintext no longer triggers a false no-op — it matches exactly what SOPS refuses to re-encrypt. Regression tests: `test_encrypt_substring_only_plaintext_still_runs_sops`, `test_has_sops_metadata_block_distinguishes_substring_from_block` (unit) and `test_encrypt_plaintext_with_enc_substring_is_actually_encrypted` (real sops).

File: src/envdrift/encryption/sops.py [hunk 3]
```diff
@@ -180,7 +180,15 @@ def _run(
...
+        # An explicit config path that does not exist is a user error: silently
+        # dropping --config would let SOPS fall back to an ambient .sops.yaml (wrong
+        # keys, exit 0 — a data-integrity hazard) instead of surfacing the typo.
+        # self._config_file is only ever set from an explicit CLI --sops-config or
+        # TOML sops_config_file (never from auto-discovery), so a missing one is
+        # always an explicit-but-wrong path.
+        if self._config_file is not None:
+            if not self._config_file.exists():
+                raise EncryptionBackendError(f"SOPS config file not found: {self._config_file}")
             cmd.extend(["--config", str(self._config_file)])
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/encryption/sops.py]: Fixed in d1dfb25. `_resolve_config_path(cwd)` now resolves a relative explicit config against the subprocess `cwd` (where sops actually runs) before the existence check, so a valid relative `--sops-config` is no longer rejected with a false `not found`. Regression tests: `test_run_relative_config_resolved_against_cwd` and `test_run_relative_config_missing_under_cwd_raises`.

File: src/envdrift/encryption/sops.py [hunk 4]
```diff
@@ -239,6 +247,26 @@ def encrypt(
...
+                existing = env_file.read_text()
+            except (OSError, UnicodeDecodeError):
+                existing = ""
+            if existing and self.has_encrypted_header(existing):
+                return EncryptionResult(
+                    success=True,
+                    message=f"{env_file} is already encrypted (no change)",
+                    file_path=env_file,
+                )
+
```
Comment (jainal09) [inline src/envdrift/encryption/sops.py]: Refactored in d1dfb25 (no suppression). `encrypt()` now delegates to `_already_encrypted_noop`, `_encrypt_key_args`, `_encrypt_output_args` and `_config_args` helpers, dropping its branch count back under threshold (the body is now a flat sequence of early returns plus a single args assembly).

File: src/envdrift/encryption/sops.py [hunk 5]
```diff
@@ -180,7 +180,15 @@ def _run(
...
-        if self._config_file and self._config_file.exists():
+        # An explicit config path that does not exist is a user error: silently
+        # dropping --config would let SOPS fall back to an ambient .sops.yaml (wrong
+        # keys, exit 0 — a data-integrity hazard) instead of surfacing the typo.
+        # self._config_file is only ever set from an explicit CLI --sops-config or
+        # TOML sops_config_file (never from auto-discovery), so a missing one is
+        # always an explicit-but-wrong path.
+        if self._config_file is not None:
+            if not self._config_file.exists():
+                raise EncryptionBackendError(f"SOPS config file not found: {self._config_file}")
```
Comment (jainal09) [inline src/envdrift/encryption/sops.py]: Refactored in d1dfb25. The `--config` build/validate logic moved out of `_run` into `_config_args` / `_resolve_config_path`, so `_run` is now a single `cmd = [binary, *config_args, *args]` plus the subprocess try/except — under threshold.

File: src/envdrift/encryption/sops.py (hunk 4 above)
Comment (jainal09) [inline src/envdrift/encryption/sops.py]: Resolved in d1dfb25. The two nested conditional blocks (the idempotency read/header check and the output-mode resolution) are extracted into `_already_encrypted_noop` and `_encrypt_output_args`, so `encrypt()` no longer has nested branching.

**Comment by jainal09** [comment]
Bot re-review update for latest head `d1dfb255`: the three SOPS findings from my earlier review look addressed.

- The idempotency no-op now uses a genuine SOPS metadata-block check instead of a bare `ENC[AES256_GCM,` substring.
- Explicit config validation runs before the idempotency no-op.
- Relative explicit config paths are resolved against the `_run(cwd=...)` context.

Local verification on the latest head: `tests/unit/test_encryption_sops.py` passed (17 tests), ruff passed on the touched SOPS files, `git diff --check` passed, and the new real-binary integration repro `test_encrypt_plaintext_with_enc_substring_is_actually_encrypted` passed. GitHub integration/test and Cubic checks were still pending when I re-polled.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>

File: src/envdrift/encryption/sops.py [hunk 6]
```diff
@@ -166,6 +166,39 @@ def _build_env(self, env: dict[str, str] | None = None) -> dict[str, str]:
...
+        ``--sops-config`` or TOML ``sops_config_file`` (never auto-discovery), so a
+        missing one is always an explicit-but-wrong path. A *relative* config path
+        is resolved against the ``cwd`` SOPS will run in (not the process cwd), so
+        validation matches how SOPS itself would locate it.
+        """
+        if self._config_file is None:
+            return None
+        config_path = self._config_file
+        if cwd is not None and not config_path.is_absolute():
+            config_path = Path(cwd) / config_path
```
[the remark this answers is not in the document]
Comment (jainal09) [reply src/envdrift/encryption/sops.py]: Fixed in 9790928. `_resolve_config_path` now returns an *absolute* path: a relative config is anchored under `cwd` and `.resolve()`d, so SOPS (running with that `cwd`) no longer re-resolves it and applies `cwd` a second time. Regression test `test_run_relative_config_and_relative_cwd_not_applied_twice` asserts the `--config` flag is absolute and points at the real file (not `cwd/cwd/config`).

**Comment by jainal09** [comment]
Bot re-review update for latest head `97909283`: the new relative-cwd SOPS config fix looks good. `_resolve_config_path()` now returns an absolute path when anchoring a relative config under `cwd`, avoiding the double-`cwd` application problem.

Local verification on this head: `tests/unit/test_encryption_sops.py` passed (19 tests). The earlier three SOPS findings remain addressed.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>

**Comment by jainal09** [comment]
Bot CI re-check for latest head `97909283`: GitHub Actions and CodeScene are passing, but `codecov/project` is still failing. `codecov/patch` is passing. The SOPS review findings remain addressed by the latest commits.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>

**Comment by jainal09** [comment]
Bot review summary for latest head `97909283`:

I re-reviewed the SOPS idempotency/config-path changes and verified a remaining API-contract regression.

Finding:
- Medium: `SOPSEncryptionBackend.encrypt()` returns success for `in_place=False` without `output_file` when the input is already encrypted. The idempotency no-op runs before `_encrypt_output_args()` validates output mode, so an already-encrypted file returns `success=True` / `already encrypted (no change)` while producing no requested output. I reproduced this with a SOPS dotenv file containing `ENC[AES256_GCM,...]`, `sops_version=...`, and `sops_mac=...`; `backend.encrypt(path, in_place=False)` returned success and only the original file existed. Please validate output mode before the no-op, or only no-op when the effective target is truly in-place, and add a regression for SOPS metadata plus `in_place=False`.

The earlier SOPS marker/config/cwd findings remain addressed. Current visible checks are not failing.

<sub>Bot comment posted by Codex via `gh` using Jainal's account.</sub>
